src/client/datascience/jupyter/kernels/kernelConnectionWrapper.ts

```diff
@@ -44,84 +44,145 @@ export class KernelConnectionWrapper implements Kernel.IKernelConnection {
     public readonly unhandledMessage = new Signal<this, IMessage<MessageType>>(this);
     public readonly anyMessage = new Signal<this, Kernel.IAnyMessageArgs>(this);
     public get serverSettings() {
-        return this.kernelConnection.serverSettings;
+        return (this.kernelConnection || this._previousKernelConnection).serverSettings;
     }
     public readonly disposed = new Signal<this, void>(this);
+    private _kernelConnection!: Kernel.IKernelConnection;
+    private _previousKernelConnection!: Kernel.IKernelConnection;
+    // private _isRestarting?: boolean;
+    private get kernelConnection(): undefined | Kernel.IKernelConnection {
+        if (this.kernel.session?.kernel === this._kernelConnection) {
+            return this._kernelConnection;
+        }
+        this.stopHandlingKernelMessages(this._kernelConnection);
+        if (this.kernel.session?.kernel) {
+            this.startHandleKernelMessages(this.kernel.session.kernel);
+            return this._kernelConnection;
+        }
+    }
+    private getKernelConnection(): Kernel.IKernelConnection {
+        if (!this.kernelConnection) {
+            throw new Error(
+                `Kernel connection is not available, status = ${this.status}, connection = ${this.connectionStatus}`
+            );
+        }
+        return this.kernelConnection;
+    }
 
-    constructor(
-        readonly kernel: IKernel,
-        private kernelConnection: Kernel.IKernelConnection,
```

Comment:
We cannot pass `kernelConnection` as it changes when restarting kernels, its not a constant